src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,320 @@
+	this.groupPermission = adminPermissionManagement.groups();
+	this.policyStore = adminPermissionManagement.authz().getStoreFactory().getPolicyStore();
+    }
+
+    private Set<String> getGroupIdsFromPolicyConfig(Map<String, String> policyConfig) {
+
+	// policy config returns a type of Map<String, String>
+	// e.g.
+	// {"groups":"[{\"id\":\"b770dfde-038b-493e-bc7f-a51a5ce86741\",\"extendChildren\":true}]"}
+	// this function parses the value of "groups" and extract the group ids into a
+	// hashset
+
+	Set<String> result = new HashSet<String>();
+	if (policyConfig.containsKey("groups")) {
+	    String regex = "[a-f0-9]{8}(-[a-f0-9]{4}){4}[a-f0-9]{8}";
+	    Matcher m = Pattern.compile("(?=(" + regex + "))").matcher(policyConfig.get("groups"));
+	    while (m.find()) {
+		result.add(m.group(1));
+	    }
+	}
+	return result;
+    }
+
+    private SortedMap<Integer, String> getPolicies(Policy permission, Set<String> groupIds) {
+
+	// get the names of the group type policies from a permission
+	// and check if a user applies to the policies by comparing the group ids
+
+	SortedMap<Integer, String> result = new TreeMap<Integer, String>();
+	Map<String, Set<String>> policyGroups = new HashMap<String, Set<String>>();
+	permission.getAssociatedPolicies().stream().filter(policy -> "group".equals(policy.getType()))
+		.forEach(policy -> policyGroups.put(policy.getName(), getGroupIdsFromPolicyConfig(policy.getConfig())));
+
+	for (String policy : policyGroups.keySet()) {
+	    Set<String> policyGroupIds = policyGroups.get(policy);
+	    // get overlap groups between policy groups and user groups
+	    policyGroupIds.retainAll(groupIds);
+	    if (!policyGroupIds.isEmpty()) {
+		// the user applies to the policy
+		String policyName = UserGroupRoles.parsePolicyName(policy)[0];
+		result.put(UserGroupRoles.POLICY_HIERARCHY.get(policyName), policyName);
+	    }
+	}
+
+	return result;
+    }
+
+    private Set<String> getPermissionScopes(Policy permission) {
+
+	// get the scopes of a permission
+
+	return permission.getScopes().stream().map(Scope::getName).collect(Collectors.toSet());
+    }
+
+    private Policy getPolicyByPermissionName(String permissionName, GroupModel group) {
+
+	switch (permissionName) {
+	case "view":
+	    return groupPermission.viewPermission(group);
+	case "manage":
+	    return groupPermission.managePermission(group);
+	case "view-members":
+	    return groupPermission.viewMembersPermission(group);
+	case "manage-members":
+	    return groupPermission.manageMembersPermission(group);
+	case "manage-membership":
+	    return groupPermission.manageMembershipPermission(group);
+	default:
+	    return null;
+	}
+    }
+
+    private Boolean isRootGroup(GroupModel group) {
+
+	// check if the group is a root group
+	// if a root group, check if fine grained permission is enabled
+	// enable fine grained admin permission if not
+
+	if (group.getParent() != null) {
+	    groupPermission.setPermissionsEnabled(group, false);
```

Comment:
I would not expect a function `isRootGroup` to modify the group as a side effect. If that's going to be done, maybe a better name is `setFineGrainedPermissions` -- it is setting to either true or false depending on what type of group it is